### ISSUE TITLE
[SYCL][SPEC] USM regarding to context

### ISSUE THREAD

**Issue by CaoZhongZ** [intel/llvm#2721]
I make statement, anything wrong please tell me.

I want to understand USM allocation regarding to "device" or "context". Consider Unified Addressing, Virtual Address Space tied to a context, consider "shared" part, real resource backing the virtual address is tied to device.

The problem is when Unified Addressing tied to multiple contexts in a single process. So the same virtual address with different context belongs to different virtual address space and are totally different entities, could point to completely different physical resources?

**Comment by bader** [comment]
According to my understanding your statement is correct.
+@jbrodman to confirm.

**Comment by jbrodman** [comment]
Yes.

Virtual addresses are not guaranteed to have any sort of consistency across different devices.

**Comment by CaoZhongZ** [comment]
I brought up this to attention because if a unique virtual address could point to at least two different context resource, when migration happens it'll be very ugly. Conceptually, if a virtual address is not enough to distinguish different computational resource, might be huge problems lying around. Isn't it contradict Turing model a process used to hold?

**Comment by jbrodman** [comment]
A program that tries to use VAs across contexts is an incorrect program.
It is not expected or required to work.

**Comment by CaoZhongZ** [comment]
Users could control that? All users to do is allocate address1 from context1, allocate address2 from context2, do works. But address1 and address2 could be the same number. right?

**Comment by jbrodman** [comment]
They could be, they could not.  The point is, it's illegal, and writing
programs on UB is a recipe for pain.  If you need to share data between
contexts, explicitly copy it.

**Comment by keryell** [comment]
I have some issues understanding how a correct SYCL USM implementation could return the same C++ user-level address for 2 different objects.
Even if there are various back-ends and contexts, at the end they have to cooperate with the OS for not fighting for the same addresses...
At least something like Linux HMM can insure that even if behind the curtain you have actually different devices in different contexts and back-ends with different distributed physical memory pages, the illusion of a unique virtual address will hold.
In the same way as if `new int` and another `new int` are returning the same value, you are in trouble (but it is still possible with out-of-memory condition, in the case of noexcept `new`).

**Comment by jbrodman** [comment]
For the shared/host allocations you're 100% right.

I think the only place where it's possible would be for device
allocations which aren't accessible on the host.

Anything accessible on the host can't have this sort of conflict or reality
would explode.

**Comment by keryell** [comment]
While it would be possible, then it prevents use cases such as:
- you have a lot of pointers to some objects allocated in some various slabs on various devices by C++ allocators and you cannot std::partition/std::sort/... the pointer set according to the various address ranges so you can the process them on the right devices.

So, if the application/architecture/OS/... is not short on address spaces, it seems clearer to have unique virtual address space mapping, even for non host-accessible memory. Even if we are talking about pointers here, the audience is not the Linux kernel developers... :-)

**Comment by CaoZhongZ** [comment]
Agree.

A Virtual Address should identify unique object in single space, access in host or not, that's a programming language basic requirement or programmers' basic heuristics. I don't know what would happen if this were broken, any pointer arithmetic or condition would be weird?

**Comment by CaoZhongZ** [comment]
Agree, we have plenty space for a clear implementation and it is then much more convenient.

Other cases like, C++ programming language required to avoid single address, two object of same type scenario. Or programmer's heuristic on pointer arithmetic, different object would have different address, etc.

**Comment by KornevNikita** [comment]
Hi! There have been no updates for at least the last 60 days, though the ticket has assignee(s).

      @jbrodman, could I ask you to take one of the following actions? :)
      - Please provide an update if you have any (or just a small comment if you don't have any yet).
      - OR mark this issue with the 'confirmed' label if you have confirmed the problem/request and our team should work on it.
      - OR close the issue if it has been resolved.
      - OR take any other suitable action.

      Thanks!

**Comment by AlexeySachkov** [comment]
Considering that USM is now a core part of SYCL 2020, any remaining questions about it should be discussed in [KhronosGroup/SYCL-Docs](https://github.com/KhronosGroup/SYCL-Docs) repo.
